eel_hole/utils.py

```diff
@@ -21,11 +46,124 @@ def rst_to_html(rst: str) -> str:
+        in_ul = False
+        in_ol = False
+
+        for line in lines:
+            if re.match(r"^\s*[-*]\s+", line):
+                # Switch from ol to ul if needed
+                if in_ol:
+                    html_lines.append("</ol>")
+                    in_ol = False
+                if not in_ul:
+                    html_lines.append("<ul>")
+                    in_ul = True
+                html_lines.append(f"<li>{line.strip()[2:]}</li>")
+
+            elif re.match(r"^\s*\d+\.\s+", line):
+                # Switch from ul to ol if needed
+                if in_ul:
+                    html_lines.append("</ul>")
+                    in_ul = False
+                if not in_ol:
+                    html_lines.append("<ol>")
+                    in_ol = True
+                html_lines.append(f"<li>{line.strip().split(None, 1)[1]}</li>")
+
+            else:
+                # Close any open lists before adding paragraph text line
+                if in_ul:
+                    html_lines.append("</ul>")
+                    in_ul = False
+                if in_ol:
+                    html_lines.append("</ol>")
+                    in_ol = False
+                html_lines.append(line)  # keep \n inside paragraph as is
+
+        # Close any lists still open at the end of the paragraph
+        if in_ul:
+            html_lines.append("</ul>")
+        if in_ol:
+            html_lines.append("</ol>")
+
+        # Join paragraph lines with newline to preserve single line breaks
+        joined_para = "\n".join(html_lines).strip()
+
+        # Wrap non-list paragraphs in <p>
+        if not (joined_para.startswith("<ul>") or joined_para.startswith("<ol>")):
+            html_parts.append(f"<p>{joined_para}</p>")
+        else:
+            html_parts.append(joined_para)
+
+    return f"<main>\n{'\n'.join(html_parts)}\n</main>"
+
+
+def clean_pudl_resource(resource: Resource, source_key: str) -> Resource:
+    """Clean up the PUDL datapackage descriptions for display.
+
+    PUDL datapackage documentation is all in RST so we use Sphinx machinery to
+    turn it into HTML that our Jinja templates understand.
+    """
+    return Resource(
+        name=resource.name,
+        description=rst_to_html(resource.description),
+        sources=[{"title": source_key}],
+        schema=Schema(
+            fields=[
+                Field(name=field.name, description=rst_to_html(field.description))
+                for field in resource.schema.fields
+            ]
+        ),
+    )
+
+
+def clean_ferc_xbrl_resource(resource: Resource, source_key: str) -> Resource:
+    """Clean up the FERC XBRL datapackage descriptions for display.
+
+    These are written in no-format plaintext, so we have some custom HTML
+    generation. Also, the table *descriptions* are useless but the table
+    *titles* (not *name* which is the canonical name of the table) are merely
+    nearly useless. So we replace the descriptions with the titles.
```

Comment:
Should we add a followup for the xbrl datapackages to move the description information into the description field?

Reply:
https://github.com/catalyst-cooperative/ferc-xbrl-extractor/issues/369